docs/developers/developer-reference/sc-contract-calls.md

```diff
@@ -1,95 +1,387 @@
 ---
 id: sc-contract-calls
-title: Smart contract to smart contract calls
+title: Smart contract calls
 ---
 
 [comment]: # (mx-abstract)
 
 ## Introduction
 
+As programmers, we are used to calling functions all the time. Smart contract calls, however, are slightly trickier. Beside code execution, there can also be transfers of EGLD and ESDT tokens involved, gas needs to be specificed, and, finally, there are different flavors of calls.
+
+Any system for composing and sending these calls needs to be very versatile, to cater to all of the programmers' various needs.
+
+It is not so important _where_ that call comes from. Very often we need to call smart contracts from other smart contracts. However, testing systems and blockchain interactors also greatly benefit from such a system.
+
+The system we present below works just as well on- and off-chain. Think of it as a general Rust framework for specifying and sending calls in any context.
+
+:::note
+The more primitive way to perform these calls is to just use the API methods directly, to serialize arguments in code and to specify endpoints as strings. But this does not give the compiler a chance to verify the correctness of the calls, and is very limited in how much we configure the call. It also normally leads to a lot of code duplication. For this reason we recommend always using contract call syntax when formatting transactions.
+:::
+
+
+---
+
+[comment]: # (mx-context-auto)
+
+## Contract calls: base
+
+Smart contract calls at the blockchain level have no notion of arity, or data types. That is, the blockchain itself does not validate the number of arguments (or results), and each of these only appear as raw binary data fields.
+
+It is the contract that keeps track of the number of arguments, and deserializes them. If a transaction has the wrong number of arguments, it is only the contract itself that will be able to complain. If the types are off, it is only during deserialization that the contract will know.
+
+The description of a smart contract's inputs is known as the [ABI](/developers/data/abi), and lives off-chain. In short, the ABI is a collection of endpoint names, with argument names and type descriptions. To be able to effectively call a smart contract, it is useful to know its ABI.
+
+We call this contract call helper a __proxy__. All it does is that it provides a typed interface to any Rust program, it takes the typed arguments and it serializes them according to the [MultiversX serialization format](/developers/data/serialization-overview).
```

Comment:
For extra clarity:

```suggestion
Knowing the ABI of a contract enables us to call it cleanly. The ABI is used to create a __proxy__ for the contract, which is a helper component that stands in for the contract in our code. All it does is that it provides a typed interface to any Rust program, it takes the typed arguments and it serializes them according to the [MultiversX serialization format](/developers/data/serialization-overview).
```

Reply:
Not completely correct. The proxy right now is not created based on the ABI.

Reply:
I rephrased it a little differently.